controller_manager/src/controller_manager.cpp

```diff
@@ -2390,21 +2390,65 @@ controller_interface::return_type ControllerManager::update(
         run_controller_at_cm_rate ? period
                                   : rclcpp::Duration::from_seconds((1.0 / controller_update_rate));
 
-      const rclcpp::Time current_time = get_clock()->now();
+      rclcpp::Time current_time;
       bool first_update_cycle = false;
+      if (get_clock()->started())
+      {
+        current_time = get_clock()->now();
+      }
+      else if (
+        time == rclcpp::Time(0, 0, this->get_node_clock_interface()->get_clock()->get_clock_type()))
+      {
+        throw std::runtime_error(
+          "No clock received, and time argument is zero. Check your controller_manager node's "
+          "clock configuration (use_sim_time parameter) and if a valid clock source is "
+          "available. Also pass a proper time argument to the update method.");
+      }
+      else
+      {
+        // this can happen with use_sim_time=true until the /clock is received
+        rclcpp::Clock clock = rclcpp::Clock();
+        RCLCPP_WARN_THROTTLE(
+          get_logger(), clock, 1000,
+          "No clock received, using time argument instead! Check your node's clock "
+          "configuration (use_sim_time parameter) and if a valid clock source is available");
+        current_time = time;
+      }
```

Comment:
The changes itself look good to me. I'm just wondering if we have to check this for every controller in the loop or only at the beginning of the update cycle.

Any thoughts @christophfroehlich @fmauch

Reply:
It would be sufficient to throw the exception once at the beginning, but you want to update the current_time in the loop for every controller right?

Reply:
Ok. I was thinking something like

`if (!get_clock()->started() && time == rclcpp::Time(0, 0, this->get_node_clock_interface()->get_clock()->get_clock_type()))`

We can merge this and we can take a look at it later at some point